### PR TITLE
[Environment, Docs] SMACv2 and docs on action masking

### PR DESCRIPTION
**Pull request by matteobettini**
Depends on #1421
To be closed: #56
To be closed: #810

SMAC integration was started in #810, this PR aims to integrate SMACv2

Smacv2 is the second version of the popular SMAC enviornment challenge, popular as one of the most used multi-agent benchmarks.

This PR is focused on integrating SMACv2 in torchRL 

For more info you can see https://github.com/oxwhirl/smacv2

This PR also introduces the documentation on how to update the action mask for environments with changing available actions.

@ordinskiy

### CODE REVIEW

File: .circleci/unittest/linux_libs/scripts_smacv2/setup_env.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+pip install pip --upgrade
+
+conda env update --file "${this_dir}/environment.yml" --prune
+
+# 5. Install StarCraft 2 with SMACv2 maps
+# SC2PATH is set in run_test.sh
+printf "* Installing StarCraft 2 and SMACv2 maps into '${root_dir}/smacv2/StarCraftII'\n"
+cd "${root_dir}"
+# TODO: discuss how we can cache it to avoid downloading ~4 GB on each run.
```
Comment (matteobettini) [inline .circleci/unittest/linux_libs/scripts_smacv2/setup_env.sh:53]: To discuss

**Review by vmoens — APPROVED** [review]
Some minor comments
We need to see how to cache the downloaded lib

File: torchrl/envs/libs/smacv2.py [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+    OneHotDiscreteTensorSpec,
+    UnboundedContinuousTensorSpec,
+)
+from torchrl.envs.common import _EnvWrapper
+
+
+_has_smacv2 = importlib.util.find_spec("smacv2") is not None
+if _has_smacv2:
+    import smacv2
+    from smacv2.env.starcraft2.maps import smac_maps
```
[the remark this answers is not in the document]
Comment (matteobettini) [reply torchrl/envs/libs/smacv2.py:24]: do you mean that is should import it in every function that needs it?

[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/envs/libs/smacv2.py]: yeah the problem is that at some point I would like to expose all the wrappers in `torchrl.envs`.
If you have installed multiple third-party libs, you'll end up with a launch time that is really high. Moreover, our multirproc objects spawns processes that all need to import all of these things, which makes our tests really really long!
TLDR: no global imports for third party unless needed elsewhere or import is too frequent
(since local import has some overhead when repeated multiple times)

File: torchrl/envs/libs/smacv2.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+                "reward": reward,
+                "done": done,
+            },
+            batch_size=(),
+            device=self.device,
+        )
+
+        return tensordict_out
+
+    def update_action_mask(self):
```
Comment (vmoens) [inline torchrl/envs/libs/smacv2.py:401]: should this be public?

File: torchrl/envs/libs/smacv2.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        mask = torch.tensor(
+            self.get_avail_actions(), dtype=torch.bool, device=self.device
+        )
+        self.action_spec.update_mask(mask)
```
[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/envs/libs/smacv2.py:405]: I mean: you can update it here, but it'll fail with vec envs
MaskedAction transform is made just for that

File: torchrl/envs/libs/smacv2.py [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+)
+from torchrl.envs.common import _EnvWrapper
+
+from torchrl.envs.utils import ACTION_MASK_ERROR
+
+
+_has_smacv2 = importlib.util.find_spec("smacv2") is not None
+
+if typing.TYPE_CHECKING and _has_smacv2:
+    import smacv2
```
Comment (vmoens) [inline torchrl/envs/libs/smacv2.py:28]: same as petting zoo, let's avoid global imports

**Review by vmoens — APPROVED** [review]
Almost there, great work!

File: .github/workflows/test-linux-smacv2.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+concurrency:
+  # Documentation suggests ${{ github.head_ref }}, but that's only available on pull_request/pull_request_target triggers, so using ${{ github.ref }}.
+  # On master, we want all builds to complete even if merging happens faster to make it easier to discover at which point something broke.
+  group: ${{ github.workflow }}-${{ github.ref == 'refs/heads/main' && format('ci-master-{0}', github.sha) || format('ci-{0}', github.ref) }}
+  cancel-in-progress: true
+
+jobs:
+  unittests:
+    if: ${{ github.event.label.name == 'Environments' }}
```
[the remark this answers is not in the document]
Comment (matteobettini) [reply .github/workflows/test-linux-smacv2.yml]: That is above no? I didn’t modify anything here just added the line w if

[the remark this answers is not in the document]
Comment (matteobettini) [reply .github/workflows/test-linux-smacv2.yml]: The thing i wonder is if this will be run on main

[the remark this answers is not in the document]
Comment (vmoens) [reply .github/workflows/test-linux-smacv2.yml]: I don't think it will
To me, the `on` block will go first and check that it's on main. Then the job `unittests` will be run but the `if` statement will always be false.
I think this condition goes under `on: pull_request`

File: .github/workflows/test-linux-smacv2.yml [hunk 7]
```diff
@@ -17,7 +17,7 @@ concurrency:
 
 jobs:
   unittests:
-    if: contains(github.event.pull_request.labels.*.name, 'Environments')
+    if: ${{ (github.event_name == 'pull_request' && contains(github.event.pull_request.labels.*.name, 'Environments')) || (github.event_name == 'push') }}
```
Comment (vmoens) [inline .github/workflows/test-linux-smacv2.yml]: if you run every time you push you can avoid checking the label name too no?

Comment (matteobettini) [reply .github/workflows/test-linux-smacv2.yml]: i am still working on it, i thought the push event is onlly triggered on the listed branches, therefore the label i checked for prs

**Comment by matteobettini** [comment]
Ok I think I have reached a current working solution.

My hypothesis is that now the tests will be run either on PR events with the proper label or on pushes to these branches
      - nightly
      - main
      - release/*

I have tested the labelling PR filtering, but ofc the fact that this will run in main is not tested

File: .github/workflows/test-linux-smacv2.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (matteobettini) [reply .github/workflows/test-linux-smacv2.yml]: i don't think conditions can be put there
